python/pyiceberg/table/__init__.py

```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
         import ray
 
         return ray.data.from_arrow(self.to_arrow())
+
+
+class _SchemaUpdate(UpdateSchema):
+    def __init__(self, schema: Schema, table: Optional[Table] = None, last_column_id: Optional[int] = None):
+        self._table = table
+        self._schema = schema
+        if last_column_id:
+            self._last_column_id = last_column_id
+        else:
+            self._last_column_id = schema.highest_field_id
+
+        self._identifier_field_names = schema.column_names
+        self._adds: Dict[int, List[NestedField]] = {}
+        self._added_name_to_id: Dict[str, int] = {}
+        self._id_to_parent: Dict[int, str] = {}
+        self._allow_incompatible_changes: bool = False
+        self._case_sensitive: bool = True
+
+    def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        self._case_sensitive = case_sensitive
+        return self
+
+    def add_column(
+        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None
+    ) -> UpdateSchema:
+        if "." in name:
+            raise ValueError(f"Cannot add column with ambiguous name: {name}")
+
+        self._internal_add_column(parent, name, True, type_var, doc)
+        return self
+
+    def add_required_column(
+        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None
+    ) -> UpdateSchema:
+        if "." in name:
+            raise ValueError(f"Cannot add column with ambiguous name: {name}")
+
+        if not self._allow_incompatible_changes:
+            raise ValueError(f"Incompatible change: cannot add required column: {name}")
+
+        self._internal_add_column(parent, name, False, type_var, doc)
+        return self
+
+    def allow_incompatible_changes(self) -> UpdateSchema:
+        self._allow_incompatible_changes = True
+        return self
+
+    def apply(self) -> Schema:
```

Comment:
Should we make this one private?
```suggestion
    def _apply(self) -> Schema:
```